File: app/src/main/java/com/turlir/abakcalc/MainPresenter.java
package com.turlir.abakcalc;

import com.turlir.calculator.Calculator;
import com.turlir.calculator.converter.Member;
import com.turlir.calculator.converter.Visual;

import java.math.BigDecimal;
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.EmptyStackException;
import java.util.List;
import java.util.Locale;
import java.util.NoSuchElementException;
import java.util.Queue;

class MainPresenter {

    private static final DecimalFormat DF = new DecimalFormat(""); // формат результата
    private static final DecimalFormatSymbols FORMAT = new DecimalFormatSymbols(Locale.getDefault());

    static {
        DF.setDecimalFormatSymbols(FORMAT);
        DF.setMaximumFractionDigits(12);
        DF.setMinimumIntegerDigits(1);
    }

    static final String SEPARATOR = String.valueOf(FORMAT.getDecimalSeparator());

    private final Calculator mCalc;

    private MainView mView;

    MainPresenter(Calculator calc) {
        mCalc = calc;
    }

    void attach(MainView view) {
        mView = view;
    }

    void detach() {
        mView = null;
    }

    void enter(String s) {
        if (s == null || s.isEmpty()) {
            mView.resetToEmpty();
            return;
        }

        try {
            BigDecimal result = calculate(s);
            showRepresentation(result);
            showNotation();

        } catch (EmptyStackException | NoSuchElementException e) {
            mView.showError(mView.context().getString(R.string.error));

            mView.setNotation(Collections.<Visual>emptyList());

        } catch (Exception e) {
            mView.showError(e.getMessage());

            mView.setNotation(Collections.<Visual>emptyList());
        }
    }

    private void showRepresentation(BigDecimal result) {
        mView.showResult(DF.format(result));
        showRepresentation();
    }

    private void showRepresentation() {
        List<Member> copy = mCalc.direct();
        List<CalculatorVisual> visual = interceptDirect(copy);
        mView.setRepresentation(visual);
    }

    private void showNotation() {
        Queue<Member> row = mCalc.translated();
        List<Visual> now = new ArrayList<>(mCalc.size());
        for (Member item : row) now.add(item.view());
        mView.setNotation(now);
    }

    void recalculate(String s) {
        if (s == null || s.isEmpty()) {
            mView.resetToEmpty();
            return;
        }

        try {
            BigDecimal result = calculate(s);
            showRepresentation(result);
        } catch (Exception ignored) {
            showRepresentation();
        }
    }

    private  BigDecimal calculate(String str) throws Exception {
        str = str.replace(SEPARATOR, ".");
        BigDecimal digit = mCalc.calc(str);
        return digit;
    }

    private static List<CalculatorVisual> interceptDirect(Collection<Member> sequence) {
        List<CalculatorVisual> views = new ArrayList<>(sequence.size());
        for (Member member : sequence) {
            Visual view = member.view();
            if (member.operand()) {
                views.add(new ValueVisualizer(view));
            } else {
                views.add(new OperatorVisualizer(view));
            }
        }
        return views;
    }

}

File: app/src/main/kotlin/com/turlir/abakcalc/MainPresenter.kt
package com.turlir.abakcalc

import com.turlir.calculator.Calculator
import com.turlir.calculator.converter.Member
import com.turlir.calculator.converter.Visual
import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.util.*

class MainPresenter(private val mCalc: Calculator) {

    val separator: String

    private val mFormat = DecimalFormat("") // формат результата

    private var mView: MainView? = null

    init {
        val setting = DecimalFormatSymbols(Locale.getDefault())
        mFormat.apply {
            decimalFormatSymbols = setting
            maximumFractionDigits = 12
            minimumIntegerDigits = 1
        }
        separator = setting.decimalSeparator.toString()
    }

    fun attach(view: MainView) {
        mView = view
    }

    fun detach() {
        mView = null
    }

    fun enter(s: String?) {

        fun enter_(it: MainView) {
            if (s == null || s.isEmpty()) {
                it.resetToEmpty()
                return
            }

            try {
                calculate(s, true)

            } catch (e: EmptyStackException) {
                it.showError(it.context.getString(R.string.error))

            } catch (e: NoSuchElementException) {
                it.showError(it.context.getString(R.string.error))

            } catch (e: Exception) {
                it.showError(e.message!!)

            } finally {
                it.setNotation(emptyList())
            }
        }

        mView?.let { enter_(it) }
    }

    fun recalculate(s: String?) {
        if (s == null || s.isEmpty()) {
            mView?.resetToEmpty()
            return
        }
        try {
            calculate(s, false)
        } catch (ignored: Exception) {
            //
        }
    }

    @Throws(Exception::class)
    private fun calculate(str: String, shouldTranslated: Boolean) {
        val correct = str.replace(separator, ".")
        val row = mCalc.sequenceExpression(correct)
        if (row != null) {
            val inline = row.inline() // calculate
            interceptDirect(inline)
            val result = mCalc.calc(inline)
            mView?.showResult(mFormat.format(result))
            if (shouldTranslated) {
                val translated = mCalc.translated()
                interceptTranslated(translated)
            }
        }
    }

    private fun interceptDirect(sequence: List<Member>) {
        val views = ArrayList<CalculatorVisual>(sequence.size)
        moveTo(views, sequence) { it: Member ->
            if (it.operand()) {
                ValueVisualizer(it.view())
            } else {
                OperatorVisualizer(it.view())
            }
        }
        mView?.setRepresentation(views)
    }

    private fun interceptTranslated(translated: Queue<Member>) {
        val now = ArrayList<Visual>(translated.size)
        moveTo(now, translated) { it.view() }
        mView?.setNotation(now)
    }

    private fun <T, V> moveTo(destination: MutableList<T>, origin: Collection<V>, map: (V)-> T) {
        origin.asSequence()
                .map { map(it) }
                .forEach { destination.add(it) }
    }
}
